File: hackerrank/problem solving/algorithms/2 - implementation/23 - beautiful days at the movies/with morfo library/beautiful_days_at_the_movies.java
// https://www.hackerrank.com/challenges/beautiful-days-at-the-movies/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;

void main() {
    int startDay = Reader.read(Integer::parseInt);
    int endDay = Reader.read(Integer::parseInt);
    int divisor = Reader.read(Integer::parseInt);
    IO.println(beautifulDays(startDay, endDay, divisor));
}

// n: range (endDay - startDay + 1)
// d: number of digits of the largest day
// T: O(n * d)
// S: O(d) extra space
int beautifulDays(int startDay, int endDay, int divisor) {
    return (int) IntStream.range(startDay, endDay + 1).filter(day -> isBeautifulDay(day, divisor)).count();
}

boolean isBeautifulDay(int day, int divisor) {
    return (day - reverseNumber(day)) % divisor == 0;
}

int reverseNumber(int number) {
    return Integer.parseInt(new StringBuilder(String.valueOf(number)).reverse().toString());
}

File: hackerrank/problem solving/algorithms/2 - implementation/23 - beautiful days at the movies/with morfo library/beautiful_days_at_the_movies.js
// https://www.hackerrank.com/challenges/beautiful-days-at-the-movies/problem?isFullScreen=true

import { read } from 'morfo';

const startDay = read(Number);
const endDay = read(Number);
const divisor = read(Number);
console.log(beautifulDays(startDay, endDay, divisor));

// n: range (endDay - startDay + 1)
// d: number of digits of the largest day
// T: O(n * d)
// S: O(d) extra space
function beautifulDays(startDay, endDay, divisor) {
  let count = 0;
  for (let day = startDay; day <= endDay; day++)
    if (isBeautifulDay(day, divisor)) count++;
  return count;
}

function isBeautifulDay(day, divisor) {
  return (day - reverseNumber(day)) % divisor === 0;
}

function reverseNumber(number) {
  return +String(number).split('').reverse().join('');
}
